### ISSUE TITLE
Default DNN backend not set correctly

### ISSUE THREAD

**Issue by CSBVision** [opencv/opencv#25112]
### System Information

OpenCV version: 4.9.0
Operating System / Platform: Windows 11
Compiler & compiler version: Visual Studio 2022

### Detailed description

We found two related, but generally independent issues. In the 4.9.0 release, the extended CMake configuration options to select a compile time default DNN backend are included, cf. #24218 and #24252. However, if `getParam_DNN_BACKEND_DEFAULT()` returns `DNN_BACKEND_INFERENCE_ENGINE` instead of `DNN_BACKEND_OPENCV` - i.e. either by using a compiled-in default or an explicit configuration option returned by `getConfigurationParameterSizeT` - calling `forward()` results in the following runtime error:
``` Python
# compiled with OPENCV_DNN_BACKEND_DEFAULT = DNN_BACKEND_INFERENCE_ENGINE
import cv2 as cv;
model = cv.dnn.readNet('path/to/model.onnx')
model.setInput(...) # use valid input here, e.g. np.zeros((...), dtype=np.float32)
model.forward()
```
> error: OpenCV(4.9.0) opencv\modules\dnn\src\legacy_backend.cpp:124: error: (-213:The function/feature is not implemented) Unknown backend identifier in function 'cv::dnn::dnn4_v20231225::detail::wrapMat'

Explicitly setting the backend before forwarding would fix the issue, however breaks the aim of the default value. We think the problem is that inside the constructor of `Net::Impl`, the following code
``` C++
// net_impl.cpp
Net::Impl::Impl()
{
    // ...
    preferableBackend = (Backend)getParam_DNN_BACKEND_DEFAULT();
    // ...
}
```
only defines the backend, but skips the required initialization implemented in `setPreferableBackend()`. Clearly, this is only noticeable if `getParam_DNN_BACKEND_DEFAULT()` returns something different than `DNN_BACKEND_OPENCV`.
Instead, the initialization should be
``` C++
// net_impl.cpp
preferableBackend = DNN_BACKEND_OPENCV;
```
and a call to `setPreferableBackend(getParam_DNN_BACKEND_DEFAULT())` is required. As the latter depends on the `Net` instance and not just on the implementation class, this cannot be fixed in `net_impl.cpp`.

To fix this, we tested to add the following code in `net.hpp`
``` C++
Net::Net()
    : impl(makePtr<Net::Impl>())
{
    setPreferableBackend(getParam_DNN_BACKEND_DEFAULT());
}
```
This works, but we don't know whether this is the best option to fix it. We can create a PR including these two modifications, please let us know whether this is desired.

***

Furthermore, we noted a related issue appearing after re-changing the backend. It can be reproduced with all builds, defining `OPENCV_DNN_BACKEND_DEFAULT` is not required. Thus, it is independent of the previous issue. We only noticed it during the same tests. It seems that after loading a model, the first `forward()` call initializes something that is required to change the backend thereafter. Here, something is broken if the first `forward()` call uses `DNN_BACKEND_INFERENCE_ENGINE`. For example, the following code

``` Python
import cv2 as cv;
model = cv.dnn.readNet('path/to/model.onnx')
model.setPreferableBackend(cv.dnn.DNN_BACKEND_INFERENCE_ENGINE) # don't call forward() before this call
model.setInput(...); model.forward()                            # works
model.setPreferableBackend(cv.dnn.DNN_BACKEND_OPENCV)           # works as well
model.setInput(...); model.forward()                            # error
```
yields the following error message:
> error: OpenCV(4.9.0) opencv\modules\dnn\src\net_impl.cpp:570: error: (-215:Assertion failed) inp.total() in function 'cv::dnn::dnn4_v20231225::Net::Impl::allocateLayers'

It works if another `forward()` call is added **before** a different backend is selected. This is particularly relevant if a non-default backend is used.

### Steps to reproduce

See description.

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [ ] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by dkurt** [comment]
Hi! Generally, OpenCV do not support backend switch after the first inference on a different backend done.

Is that critical to switch backend in your application? Because it's possible to just call `readNet` again and keep different `net` for different backends.

**Comment by CSBVision** [comment]
No, it's not required in any case to change the backend after initializing the model, i.e. after the first inference, it should remain fixed. We only noticed this bug during the same test / debugging analysis. If this is not a bug but the expected behavior, we are fine with that too. What's most important is to correctly set the predefined default (either compiled-in or by a configuration parameter).

**Comment by CSBVision** [comment]
After the 4.10.0 release, we investigated the situation again. Unluckily, still something is broken. We used OpenVINO 2023.3.0 for our tests (there is another issue with OpenVINO's 2024 release, cf #25199) and the observations are:
1. If there is _no_ explicit default backend returned by `getParam_DNN_BACKEND_DEFAULT()` (either compiled-in or using an environment variable), selecting the backend is possible as long as `forward()` was not called so far. Selecting the backend after `forward()` was called results in the error `(-215:Assertion failed) inp.total()`. This is expected as the backend can only be selected before the first inference.

2. However, if `getParam_DNN_BACKEND_DEFAULT()` returns `2` (i.e. `DNN_BACKEND_INFERENCE_ENGINE`), the backend cannot be changed to something else, even before the first inference. Sample code:

```
import os; os.environ['OPENCV_DNN_BACKEND_DEFAULT'] = '2' # DNN_BACKEND_INFERENCE_ENGINE
import cv2 as cv
model = cv.dnn.readNet('path/to/model.onnx')
model.setPreferableBackend(cv.dnn.DNN_BACKEND_OPENCV) # skip this line and it works...
model.setInput(...); model.forward()
```

> OpenCV(4.10.0) opencv\modules\dnn\src\net.cpp:80: error: (-215:Assertion failed) !empty() in function 'cv::dnn::dnn4_v20240521::Net::forward'

It works as expected if the backend is not changed, i.e. both model and input are valid. Therefore, we think there still needs something to be fixed in advance. Any ideas what's causing the error?

**Comment by CSBVision** [comment]
After clarifying that a backend switch after `model.forward()` is not supported in general (even though our experiments conclude that it seems to work fine in most cases), we tried to analyze what's left of this issue. We used the following test code 
```Python
# define the default backend before importing cv2 as one of the following options
import os
os.environ['OPENCV_DNN_BACKEND_DEFAULT'] = '2' # DNN_BACKEND_INFERENCE_ENGINE
os.environ['OPENCV_DNN_BACKEND_DEFAULT'] = '3' # DNN_BACKEND_OPENCV
os.environ['OPENCV_DNN_BACKEND_DEFAULT'] = '5' # DNN_BACKEND_CUDA

# setup model and zero image
import numpy as np
import cv2 as cv
model = cv.dnn.readNet('image_classification_mobilenetv2_2022apr.onnx')
img = np.zeros((1,3,224,224), dtype=np.float32)

# optionally select the backend (skip to use the predefined default)
model.setPreferableBackend(cv.dnn.DNN_BACKEND_INFERENCE_ENGINE)
model.setPreferableBackend(cv.dnn.DNN_BACKEND_OPENCV)
model.setPreferableBackend(cv.dnn.DNN_BACKEND_CUDA)

# forward zero image
model.setInput(img)
model.forward()
```
where we noted the following two remaining problems:

1. Using `DNN_BACKEND_CUDA` - either as default or explicitly - requires a call to `model.setPreferableTarget(cv.dnn.DNN_TARGET_CUDA)`, too. Based on the design of splitting backend and target, this is clearly the expected behavior. However, from a user perspective of defining a system-wide / compiled-in default, it is at least slightly unexpected having to explicitly set a target in advance to avoid runtime errors. We propose to resolve to a default target using
```C++
cv::dnn::Target defaultTarget(cv::dnn::Backend backend)
{
    Target target = DNN_TARGET_CPU;
    if (backend == DNN_BACKEND_CUDA /* ||
        other backends can be added here, too */)
        target = DNN_TARGET_CUDA;
    return target;
}
```
and add a call to `setPreferableTarget(defaultTarget(backendId));` in `setPreferableBackend()`. But maybe there are better solutions, therefore we propose this feature first. Furthermore, this function can easily be extended for all further backends that require a non-CPU default. Please let us know whether you agree with that and whether we should prepare a PR for this.

---

2. Using `os.environ['OPENCV_DNN_BACKEND_DEFAULT'] = '2' # DNN_BACKEND_INFERENCE_ENGINE` as a default seems to break any option to use a different backend. For example, the following code
```Python
model.setPreferableBackend(cv.dnn.DNN_BACKEND_OPENCV)
model.setInput(img)
model.forward()
```
fails with the following error:

> error: OpenCV(4.10.0) opencv\modules\dnn\src\net.cpp:80: error: (-215:Assertion failed) !empty() in function 'cv::dnn::dnn4_v20240521::Net::forward'

Still, without a default using `os.environ`, the backend can be replaced **even after inference** as long as we use `model.setInput()` only once **before changing the backend**:
```
model.setInput(img); # after this, changing the backend always seems to work fine

model.setPreferableBackend(cv.dnn.DNN_BACKEND_INFERENCE_ENGINE)
model.setInput(img)
model.forward()

 # even changing the backend after inference works:
model.setPreferableBackend(cv.dnn.DNN_BACKEND_OPENCV)
model.setInput(img)
model.forward()
```
We conclude that the initialization required at backend changing depends on the initialization performed after the first `setInput()` (e.g. allocating buffers of the appropriate sizes?), and somehow this is one-way only if OpenVINO is selected as backend _before_ this call. We don't know whether there is an easy fix but maybe somebody knowing the code in more detail can propose a good solution based on these insights.
